quickwit/quickwit-serve/src/index_api/rest_handler.rs

```diff
@@ -554,36 +546,74 @@ fn update_index_handler(
     put,
     tag = "Indexes",
     path = "/indexes/{index_id}",
-    request_body = IndexUpdates,
+    request_body = VersionedIndexConfig,
     responses(
-        (status = 200, description = "Successfully updated the index configuration.")
+        (status = 200, description = "Successfully updated the index configuration.", body = VersionedIndexMetadata)
     ),
     params(
         ("index_id" = String, Path, description = "The index ID to update."),
     )
 )]
 /// Updates an existing index.
 ///
-/// This endpoint has PUT semantics, which means that all the updatable fields of the index
-/// configuration are replaced by the values specified in the request. In particular, omitting an
-/// optional field like `retention_policy` will delete the associated configuration.
+/// This endpoint follows PUT semantics, which means that all the fields of the
+/// current configuration are replaced by the values specified in this request
+/// or the associated defaults. In particular if the field is optional (e.g
+/// `retention_policy`), omitting it will delete the associated configuration.
+/// If the new configuration file contains updates that cannot be applied, the
+/// request fails and none of the updates are applied.
 async fn update_index(
     index_id: IndexId,
-    request: IndexUpdates,
+    config_format: ConfigFormat,
+    index_config_bytes: Bytes,
     mut metastore: MetastoreServiceClient,
 ) -> Result<IndexMetadata, IndexServiceError> {
     info!(index_id = %index_id, "update-index");
+
     let index_metadata_request = IndexMetadataRequest::for_index_id(index_id.to_string());
-    let index_uid: IndexUid = metastore
+    let current_index_metadata = metastore
         .index_metadata(index_metadata_request)
         .await?
-        .deserialize_index_metadata()?
-        .index_uid;
+        .deserialize_index_metadata()?;
+    let index_uid = current_index_metadata.index_uid.clone();
+    let current_index_config = current_index_metadata.into_index_config();
+
+    let new_index_config = quickwit_config::load_index_config_from_user_config(
+        config_format,
+        &index_config_bytes,
+        &current_index_config
+            .index_uri
+            .parent()
+            .context("Unexpected index_uri format on current configuration")
+            .map_err(IndexServiceError::InvalidConfig)?,
+    )
+    .map_err(IndexServiceError::InvalidConfig)?;
+
+    if new_index_config.index_id != index_id {
```

Comment:
The validation/diffing logic is critical. Take it out in a proper function and unit test it.

Reply:
moved it to `quickwit-config`  as most of the validation is happening there